CryptoStream Memory-based ReadAsync/WriteAsync overrides

NewellClark opened the pull request:
While working on [Override Stream ReadAsync/WriteAsync](https://github.com/dotnet/runtime/issues/33789), I found that [CryptoStream](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Security.Cryptography.Primitives/src/System/Security/Cryptography/CryptoStream.cs) is missing the memory-based async overrides.

I was able to implement the memory-based `ReadAsync` without issue. However, `WriteAsync` has a problem because it calls into `TransformBlock` on [ICryptoTransform](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Security.Cryptography.Primitives/src/System/Security/Cryptography/ICryptoTransform.cs), which is a public interface. Ouch. 

I went ahead and implemented `WriteAsync`, and where calls to `ICryptoTransform.TransformBlock` were made, I check if the underlying `ReadOnlyMemory` is actually an array skip the copy when it is.

I don't suppose it's feasible to add span-based overloads to `ICryptoTransform` as default interface implementations? I'd be willing to write up the API proposal in the case that it is.

NewellClark committed:
Memorify CryptoStream.ReadAsync

NewellClark committed:
Memorify CryptoStream.WriteAsync

NewellClark commented:
I don't understand why it's saying "new-api-needs-documentation". I added the doc-comments to `ReadAsync` and `WriteAsync` (using `///<inheritdoc/>`). Am I supposed to put documentation comments elsewhere than in the source code?

vcsjones commented:
> I don't suppose it's feasible to add span-based overloads to ICryptoTransform as default interface implementations?

This was discussed in #38764 and was ultimately determined DIMs on this interface were too risky.

vcsjones commented on src/libraries/System.Security.Cryptography.Primitives/src/System/Security/Cryptography/CryptoStream.cs:
```diff
@@ -630,12 +635,31 @@ private async ValueTask WriteAsyncCore(byte[] buffer, int offset, int count, Can
+                    var rentedBuffer = ArrayPool<byte>.Shared.Rent(inputBuffer.Length);
+                    inputBuffer.CopyTo(rentedBuffer);
+                    int result = transform.TransformBlock(rentedBuffer, 0, inputBuffer.Length, outputBuffer, outputOffset);
+                    ArrayPool<byte>.Shared.Return(rentedBuffer);
```
The `rentedBuffer` contains sensitive data and is being returned to a shared pool. We should clear the array before returning it to the pool so that when some other API rents it, it isn’t populated with the sensitive data. See other usages of `CryptographicOperations.ZeroMemory` in this class for an example.

NewellClark replied:
Fixed.

stephentoub replied:
If this is a real concern, CryptoStream should probably override CopyTo/CopyToAsync as well, to clear the temporary buffer that might be used as part of the copy.

NewellClark replied:
I can do that. I'll implement all suggested changes once I figure out why it's failing CI. It passed when I ran the tests on my machine. I must be doing something wrong.

NewellClark committed:
Zeroed memory in rented array

GrabYourPitchforks commented on src/libraries/System.Security.Cryptography.Primitives/src/System/Security/Cryptography/CryptoStream.cs:
```diff
@@ -630,12 +635,43 @@ private async ValueTask WriteAsyncCore(byte[] buffer, int offset, int count, Can
+
+                        return result;
+                    }
+                    catch
```
The catch block here is unnecessary. Instead, consider restructuring the outer try block as the following pseudocode:

```cs
byte[] rented = ArrayPool.Rent();
try
{
    input.CopyTo(rented);
    Transform(from: rented, to: output);
}
finally
{
    ZeroMem(rented);
}
ArrayPool.Return(rented);
```

By putting a finally block around the code where the rented buffer contains potentially sensitive data, we can ensure that the whole thing is zeroed out whether the operation completes successfully or fails.

We could also consider pinning the temporary buffer as part of this operation to provide further protection against copies of the data being made.

Note to other reviewers: Should we skip the array pool entirely and instead use pre-pinned arrays? If an adversary can force the crypto transform to fail, they can force the application to abandon a bunch of Gen2 arrays, which could lead to perf degradation.

stephentoub commented on src/libraries/System.Security.Cryptography.Primitives/src/System/Security/Cryptography/CryptoStream.cs:
```diff
@@ -202,7 +203,7 @@ public override void SetLength(long value)
         {
             CheckReadArguments(buffer, offset, count);
-            return ReadAsyncInternal(buffer, offset, count, cancellationToken);
+            return ReadAsync(buffer.AsMemory(offset, count), cancellationToken).AsTask();
```
If we're concerned about that, both overloads of ReadAsync can delegate to a non-virtual ReadAsyncCore.

I've never seen a type derived from CryptoStream, though.  I'm sure someone somewhere has done it, but I don't think we need to be too concerned (famous last words).  If we were actually concerned, we'd potentially want to take it a step further and have the ReadAsync memory overload check whether the type was derived or not, delegating to the base implementation if it was, in case someone had overridden ReadAsync(byte[], ...) to do something special, in which case it would be a breaking change for ReadAsync(Memory, ...) to not use it.

NewellClark replied:
Would we want to do reflection to see if the derived type has actually overridden ReadAsync(byte[]...) so we can skip the memory copy if the derived type when possible?

stephentoub replied:
> Would we want to do reflection

No

NewellClark committed:
Update src/libraries/System.Security.Cryptography.Primitives/src/System/Security/Cryptography/CryptoStream.cs

Use explicit type

Co-authored-by: Stephen Toub <stoub@microsoft.com>

NewellClark commented:
I'm going to try and figure out why it's failing CI, and once I fix that, I'll implement all suggested changes so this can be merged.

NewellClark committed:
Fix exception behavior

Unintentional breaking change with exceptions that was blocking CI. Note to self: remember to rebuild before running unit tests.

NewellClark committed:
[no merge] Pinned array, used finally block

Still need to override CopyTo/CopyToAsync. I just want to make sure that CI will pass with changes so far.

bartonjs commented on src/libraries/System.Security.Cryptography.Primitives/src/System/Security/Cryptography/CryptoStream.cs:
```diff
@@ -544,8 +565,8 @@ private async ValueTask WriteAsyncCore(byte[] buffer, int offset, int count, Can
                     // and return
-                    Buffer.BlockCopy(buffer, offset, _inputBuffer, _inputBufferIndex, count);
-                    _inputBufferIndex += count;
+                    buffer.Slice(0, buffer.Length).CopyTo(_inputBuffer.AsMemory(_inputBufferIndex));
```
```suggestion
                    buffer.CopyTo(_inputBuffer.AsMemory(_inputBufferIndex));
```

NewellClark replied:
Fixed in latest commit.

bartonjs commented on src/libraries/System.Security.Cryptography.Primitives/src/System/Security/Cryptography/CryptoStream.cs:
```diff
@@ -630,12 +650,44 @@ private async ValueTask WriteAsyncCore(byte[] buffer, int offset, int count, Can
+                }
+                else
+                {
+                    var rentedBuffer = ArrayPool<byte>.Shared.Rent(inputBuffer.Length);
```
`var` shouldn't be used here, and please add the comment explaining why this isn't using CryptoPool:

```suggestion
                    // Use ArrayPool.Shared instead of CryptoPool because the array is passed out.
                    byte[] rentedBuffer = ArrayPool<byte>.Shared.Rent(inputBuffer.Length);
```

bartonjs commented on src/libraries/System.Security.Cryptography.Primitives/src/System/Security/Cryptography/CryptoStream.cs:
```diff
@@ -630,12 +650,44 @@ private async ValueTask WriteAsyncCore(byte[] buffer, int offset, int count, Can
+                    }
+                    finally
+                    {
+                        CryptographicOperations.ZeroMemory(rentedBuffer.AsSpan(0, inputBuffer.Length));
```
If the array is being pinned for the operation that's presumably so that it can be populated and cleared without GC compaction applying, so I'd expect the clear to be inside the pin.

NewellClark committed:
Added CopyTo/CopyToAsync overrides

- Applied fixes suggested by @bartonjs
- Added overrides for CopyTo/CopyToAsync to ensure any temporary buffers are properly cleared.

NewellClark commented:
@bartonjs I've applied the changes you've suggested
@stephentoub I've implemented overrides of `CopyTo` and `CopyToAsync` to ensure temporary buffers are zeroed out.

NewellClark committed:
Merge remote-tracking branch 'dotnet/master' into memorify-CryptoStream

NewellClark committed:
Merge remote-tracking branch 'dotnet/master' into memorify-CryptoStream

NewellClark commented:
Just hit issue #32805. Rerunning.

NewellClark committed:
Use fully-qualified name in ref

NewellClark committed:
Merge remote-tracking branch 'dotnet/master' into memorify-CryptoStream

bartonjs commented on src/libraries/System.Security.Cryptography.Primitives/src/System/Security/Cryptography/CryptoStream.cs:
```diff
@@ -630,12 +651,126 @@ private async ValueTask WriteAsyncCore(byte[] buffer, int offset, int count, Can
             }
             return;
+
+            unsafe static int TransformBlock(ICryptoTransform transform, ReadOnlyMemory<byte> inputBuffer, byte[] outputBuffer, int outputOffset)
```
```suggestion
            private unsafe static int TransformBlock(ICryptoTransform transform, ReadOnlyMemory<byte> inputBuffer, byte[] outputBuffer, int outputOffset)
```

bartonjs replied:
Whoops.  That's a local function, not a member function.

bartonjs committed:
Update src/libraries/System.Security.Cryptography.Primitives/src/System/Security/Cryptography/CryptoStream.cs

bartonjs commented on src/libraries/System.Security.Cryptography.Primitives/src/System/Security/Cryptography/CryptoStream.cs:
```diff
@@ -630,12 +651,126 @@ private async ValueTask WriteAsyncCore(byte[] buffer, int offset, int count, Can
             }
             return;
+
+            private unsafe static int TransformBlock(ICryptoTransform transform, ReadOnlyMemory<byte> inputBuffer, byte[] outputBuffer, int outputOffset)
```
```suggestion
            unsafe static int TransformBlock(ICryptoTransform transform, ReadOnlyMemory<byte> inputBuffer, byte[] outputBuffer, int outputOffset)
```

bartonjs committed:
Undo bad edit

bartonjs commented on src/libraries/System.Security.Cryptography.Primitives/ref/System.Security.Cryptography.Primitives.cs:
```diff
@@ -4,6 +4,8 @@
 // Changes to this file must follow the https://aka.ms/api-review process.
 // ------------------------------------------------------------------------------
 
+using System.IO;
+
 namespace System.Security.Cryptography
```
```suggestion
namespace System.Security.Cryptography
```

bartonjs committed:
Update src/libraries/System.Security.Cryptography.Primitives/ref/System.Security.Cryptography.Primitives.cs
